### ISSUE TITLE
Doubt regarding loss function in YOLOv5

### ISSUE THREAD

**Issue by ashishsethi20** [ultralytics/yolov5#5570]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

 I have been trying to modify to incorporate the contrastive loss in the loss function and from the past week, I'm not able to understand the build_targets function in the loss.py file. can anybody please explain me why we are using this?

```
def build_targets(self, p, targets):
		# Build targets for compute_loss(), input targets(image,class,x,y,w,h)
		na, nt = self.na, targets.shape[0]  # number of anchors, targets
		tcls, tbox, indices, anch = [], [], [], []
		gain = torch.ones(7, device=targets.device)  # normalized to gridspace gain
		ai = torch.arange(na, device=targets.device).float().view(na, 1).repeat(1, nt)  # same as .repeat_interleave(nt)
		targets = torch.cat((targets.repeat(na, 1, 1), ai[:, :, None]), 2)  # append anchor indices

		g = 0.5  # bias
		off = torch.tensor([[0, 0],
							[1, 0], [0, 1], [-1, 0], [0, -1],  # j,k,l,m
							# [1, 1], [1, -1], [-1, 1], [-1, -1],  # jk,jm,lk,lm
							], device=targets.device).float() * g  # offsets

		for i in range(self.nl):
			anchors = self.anchors[i]
			gain[2:6] = torch.tensor(p[i].shape)[[3, 2, 3, 2]]  # xyxy gain

			# Match targets to anchors
			t = targets * gain
			if nt:
				# Matches
				r = t[:, :, 4:6] / anchors[:, None]  # wh ratio
				j = torch.max(r, 1. / r).max(2)[0] < self.hyp['anchor_t']  # compare
				# j = wh_iou(anchors, t[:, 4:6]) > model.hyp['iou_t']  # iou(3,n)=wh_iou(anchors(3,2), gwh(n,2))
				t = t[j]  # filter

				# Offsets
				gxy = t[:, 2:4]  # grid xy
				gxi = gain[[2, 3]] - gxy  # inverse
				j, k = ((gxy % 1. < g) & (gxy > 1.)).T
				l, m = ((gxi % 1. < g) & (gxi > 1.)).T
				j = torch.stack((torch.ones_like(j), j, k, l, m))
				t = t.repeat((5, 1, 1))[j]
				offsets = (torch.zeros_like(gxy)[None] + off[:, None])[j]
			else:
				t = targets[0]
				offsets = 0

			# Define
			b, c = t[:, :2].long().T  # image, class
			gxy = t[:, 2:4]  # grid xy
			gwh = t[:, 4:6]  # grid wh
			gij = (gxy - offsets).long()
			gi, gj = gij.T  # grid xy indices

			# Append
			a = t[:, 6].long()  # anchor indices
			indices.append((b, a, gj.clamp_(0, gain[3] - 1), gi.clamp_(0, gain[2] - 1)))  # image, anchor, grid indices
			tbox.append(torch.cat((gxy - gij, gwh), 1))  # box
			anch.append(anchors[a])  # anchors
			tcls.append(c)  # class

		return tcls, tbox, indices, anch

```

**Comment by glenn-jocher** [comment]
@ashishsethi20 build_targets() is building arrays of target anchors, it's deciding which annotations to assign to which anchors.
